undertow/WFLY-10614_Elytron_SecurityDomain_Referencing.adoc

```diff
@@ -0,0 +1,98 @@
+ 
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-10614[WFLY-10614]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1024[EAP7-1024]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Affected Projects or Components
+
+* WildFly Core
+** Default WildFly Elytron Configuration
+
+* WildFly
+** Undertow Subsystem
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The `application-security-domain` resource in the Undertow subsystem will support a new `security-domain` attribute as an alternative to the existing `http-authentication-factory` attribute - these two attributes will be defined as 'alternatives' to ensure mutual-exclusivity.
+
+Where a deployment matches against an `application-security-domain` mapping, if that mapping references a `security-domain` then provided the web.xml is configured to use the mechanisms referenced in the servlet specification (BASIC, CLIENT_CERT, DIGEST, or FORM) then these will still be enabled for the application without requiring a `http-authentication-factory` reference.
+
+Where a deployment does not specify any authentication methods then the SecurityDomain from the `application-security-domain` resource will still be associated with the deployment however no authentication mechanisms will be enabled.  The associated SecurityDomain could come either from a `security-domain` reference or it could be the SecurityDomain from a referenced `http-authentication-factory`.  Within the deployment it will then be possible to use the login and logout methods on the HttpServletRequest to establish the identity for the current request and log out the identity.
+
+The existing `http-authentication-factory` will remain in the model without deprecation as a fully supported form of configuration, after this enhancement has been implemented it will still be appropriate to reference a `http-authentication-factory` for a number of reasons, examples include: -
+
+* Using custom HTTP authentication mechanisms.
+* Using mechanisms other than the four described in the Servlet specification.
+* Where the configuration in the deployment is to be overridden.
+* Where additional configuration needs to be defined for the mechanisms such as principal transformers, credential factories, mechanism realms etc...
+
+=== Nice-to-Have Requirements
+
+The default configuration will be cleaned up to remove the `/subsystem=elytron/http-authentication-factory=application-http-authentication` resource as by default this is only used for BASIC and FORM authentication anyway.
```

Comment:
I think we can't remove it in this release, but only deprecate. Removal can break users migrating from previous releases.

If user migrate app + standalone.xml it will be fine. But if user has app + CLI scripts (which does not create application-http-authentication because it was by default there) migration will fail

Reply:
@jmesnil / @bstansberry Where are we currently on the level of compatibility that must be maintained in relation to the default configuration.  I believe in the past we said it is actual configuration that must be moved from version to version not scripts.

Users that have written scripts that define their own hierarchy of resources will be unaffected, in this case it is only going to be the ones that refer to the existing named resource that would be affected.

We could consider adding some kind of 'restore-*' cli script that can add it back for the occasional few who did find a way to depend on it - or as a single resource adding it back themselves is only one command.

Outside of this we have plans to switch the default security implementation around WildFly 15 and then remove legacy entirely after that at a later point - both of these will impact the default configuration and any scripts built off the default configuration.

Reply:
It's never been a requirement that users can take scripts written to update a standard config from one minor release, execute them against the standard config from another minor release, and be guaranteed that all will work fine.  I think users trying to do that are engaged in an anti-pattern, since our standard configs are a way we introduce latest features and best practices, and blindly jumping to those is a mistake. We support running old configs and introduced the config migration tool both as ways to allow users to move to our latest and greatest in a controlled manner.

Reply:
@aloubyansky This is an interesting discussion in the context of galleon.